src/v/cluster/controller.cc

```diff
@@ -548,59 +574,45 @@ ss::future<> controller::create_cluster() {
  * to it, and before we have started communicating with peers.
  */
 ss::future<> controller::cluster_creation_hook(cluster_discovery& discovery) {
-    auto is_cluster_founder = co_await discovery.is_cluster_founder();
-    if (!is_cluster_founder) {
-        co_return;
-    }
+    const bool controller_log_exists = _raft0->last_visible_index()
+                                       > model::offset{};
 
-    if (config::node().empty_seed_starts_cluster()) {
-        // This is the legacy empty-seed-starts-cluster mode
+    if (co_await discovery.is_cluster_founder()) {
         vassert(
-          config::node().seed_servers().empty(), "We are not on the root node");
-
-        if (
-          _raft0->last_visible_index() > model::offset{}
-          || _raft0->config().brokers().size() > 1) {
-            // The controller log has already been written to
-            co_return;
+          !controller_log_exists,
+          "Can't be a cluster founder with controller log present");
+        if (config::node().empty_seed_starts_cluster()) {
+            vassert(
+              config::node().seed_servers().empty(),
+              "Cluster founder is not the root node");
+            vassert(
+              _raft0->config().brokers().size() == 1,
+              "Root is a cluster founder but joiners are in the cluster alrdy");
         }
 
-        // Internal RPC does not start until after controller startup
-        // is complete (we are called during controller startup), so
-        // it is guaranteed that if we were single node/empty controller
-        // log at start of this function, we will still be in that state
-        // here.  The wait for leadership is really just a wait for the
-        // consensus object to finish writing its last_voted_for from
-        // its self-vote.
-        while (!_raft0->is_leader()) {
-            co_await ss::sleep(100ms);
-        }
-
-        co_return co_await create_cluster();
-    }
-
-    // The new cluster creation is done by the leader of raft0 consisting of
-    // seed servers, once elected
-    while (!_storage.local().get_cluster_uuid() && !_raft0->is_leader()) {
-        co_await ss::sleep(100ms);
+        co_return co_await create_cluster(create_cluster_mode::bootstrap);
     }
 
-    if (_storage.local().get_cluster_uuid()) {
+    if (controller_log_exists && !_storage.local().get_cluster_uuid()) {
+        // No cluster UUID while controller log present - we are upgrading
+        // an existing cluster
         vlog(
           clusterlog.info,
-          "Cluster is {}",
-          *_storage.local().get_cluster_uuid());
-        co_return;
+          "Cluster will be upgraded once all nodes are upgraded. "
+          "Root driven bootstrap: {}",
+          config::node().empty_seed_starts_cluster());
```

Comment:
nit: I'm not sure this config isn't relevant or particularly useful for an admin, is it? Remove it from the log message?

Reply:
I think it may be important for field diagnostics. This value can be retrieved from other parts of the log, but since this only happens once in a cluster lifetime I don't think such info duplication is much of an evil.

Reply:
I'm curious how it'd be useful in reasoning about this background task? I agree it's useful to know about the mode in general, but why here? And why aren't the several other messages about bootstrapping sufficient?

Regardless, my point is whether we're in root-driven mode or seeds-driven mode doesn't really matter for the background task, and isn't even relevant to "Cluster will be upgraded once all nodes are upgraded." I worry having this be a part of the same message only serves to confuse readers who are less familiar with the implementation into thinking they're related, but they're not.

If you still feel strongly that we need additional context, also consider:
- having this be a separate log message from the message about upgrading,
- instead having this print _either_ "Using seeds-driven bootstrap" or "Using root-driven bootstrap", and/or
- just referring to the `redpanda.empty_seed_starts_cluster` property directly

Reply:
removed anyway